### PR TITLE
[REVIEW] install cmake config file with RMM

### PR DESCRIPTION
**Pull request by germasch**
supersedes #462 -- a lot of cmake stuff has changed since, so it makes sense to do this over.

This PR first switches to using `GNUInstallDirs` to determine paths where to write includes / libraries. This doesn't change existing behavior, but it does make it easier for, e.g.,  distros to set custom install paths.

The main thing, though, is that it adds `rmm-config.cmake` and friends that allow for easier consumption of RMM in downstream projects, ie.,

```cmake
find_package(rmm 0.16 REQUIRED)
# ...
target_link_libraries(my_app rmm::rmm)
```

This PR does not yet handle the thirdparty dependency on thrust -- I'll add that after #579 is resolved. It does, however, handle the dependency on spdlog, which serves as an example on how thirdparty are handled:

Since RMM is now using CPM, there are two cases, depending on whether spdlog is present. In either case, the `rmm::rmm` target will records its dependency on spdlog (as well as cudart and thrust), which has to be resolved when pulling in the package via `find_package` as above.

1. spdlog is found as a locally installed package (e.g., installed through conda). In this case, the spdlog dependency is resolved through a `find_dependency(spdlog 1.17.0)` which happens automatically inside of the `find_package()` call above. This assumes that spdlog is installed on the system where the installed RMM is used (and will error if no spdlog can be found).

2. spdlog was not locally installed, so RMM pulled in spdlog via `FetchContent`. In this case, since a downstream user needs to have spdlog in order to (fully) use RMM, spdlog is installed together with RMM on `make install`. The downstream app will end up using RMM and its bundled spdlog.

This does change existing behavior in the 2nd case in that spdlog is built and installed together with RMM. It should not break anything, but takes some additional build time. Hitting the 2nd case should be rare, though -- when using, e.g., conda, spdlog will be available and one will hit case 1, where existing behavior does not change (except for installing a few additional cmake config files).

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by germasch** [comment]
Could someone say "ok to test"? ;)

**Comment by kkraus14** [comment]
Ok to test

**Comment by jrhemstad** [comment]
Can we look into using [`PackageProject`](https://github.com/TheLartians/PackageProject.cmake) here?

**Comment by germasch** [comment]
> Can we look into using [`PackageProject`](https://github.com/TheLartians/PackageProject.cmake) here?

I looked at it. It doesn't quite work, once Thrust is added, since it assumes it can just do a bunch of `find_dependency(...)` calls, but we need to have a `thrust_create_target()`, too. I don't think it has a way to add `cmake_minimum_required()` into the config file, either, which is advisable since we're relying on `find_package(CUDAToolkit)`, which is relatively recent.

We could possibly submit PRs to `PackageProject` to make it more customizable so that it could fit the needs, but in the end I'm not sure we end up with something simpler.

**Review by kkraus14 — APPROVED** [review]

File: CMakeLists.txt
Comment (kkraus14) [inline CMakeLists.txt:156]:
```suggestion
  ${RMM_BINARY_DIR}/rmm-config.cmake
```

File: CMakeLists.txt
Comment (kkraus14) [inline CMakeLists.txt:161]:
```suggestion
  ${RMM_BINARY_DIR}/rmm-config-version.cmake
```

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: README.md
Comment (harrism) [inline README.md:138]:
```suggestion
GoogleBenchmark. In general you won't have to worry about it. If CMake
finds an appropriate version on your system, it uses it (you can
help it along by setting `CMAKE_PREFIX_PATH` to point to the
installed location). Otherwise those dependencies will be downloaded as
part of the build.
```

File: README.md
Comment (harrism) [inline README.md:143]:
```suggestion
If you frequently start new builds from scratch, consider setting the
environment variable `CPM_SOURCE_CACHE` to an external download
directory to avoid repeated downloads of the third-party dependencies.
```

File: README.md
Comment (harrism) [inline README.md:149]:
```suggestion
The installed RMM library provides a set of config files that makes it easy to
integrate RMM into your own CMake project. In your `CMakeLists.txt`, just add
```

File: README.md
Comment (harrism) [inline README.md:159]:
```suggestion
Since RMM is a header-only library, this does not actually link RMM,
but it makes the headers available and pulls in transitive dependencies.
If RMM is not installed in a default location, use
```

File: README.md
Comment (harrism) [inline README.md:163]:
```suggestion
One of RMM's dependencies is the Thrust library, so the above
automatically pulls in `Thrust` by means of a dependency on the
```

File: README.md
Comment (harrism) [inline README.md:166]:
```suggestion
`THRUST_HOST_SYSTEM` and `THRUST_DEVICE_SYSTEM`; see
```

**Review by harrism — CHANGES_REQUESTED** [review]
Looks great. Just suggesting some streamlining of the (already very nice) docs you added, such as  converting future tense to present tense for clarity.

**Comment by germasch** [comment]
Oops, I just found this comment I wrote in github which I wrote a couple of days ago, but apparently never actually submitted.
This PR should not be merged before settling this.

So it's time to talk about the caveats.

First of all, the potential for breaking something that currently works is fairly low. Essentially, this PR only adds some new cmake config files that are installed alongside rmm's headers.

But there are some complications related to thrust:

Currently, the installed target does depend on `spdlog::spdlog_header_only`, `CUDA::cudart` and `rmm::Thrust`. The first two are pulled in automatically when doing `find_package(rmm)`, but not yet `rmm::Thrust`.

So that's why it's WIP, it's only actually usable by a downstream project if the downstream project hacks around this limitation.

I think there are two options to handle this:
1. drop the explicit dependency on `rmm::Thrust` for now.
2. pull in Thrust 1.10.0 via `find_dependency`.

Option 1 would make things work without trouble for the typical downstream project. They'd likely end up using the thrust installed with their CUDA toolkit, which is probably good enough for them. If they do care about a newer thrust, they'd pull that in explicitly, and it'd get used.

Option 2 is the more correct way of doing things, but it also makes it harder to actually use rmm via `find_package(rmm)`. There are a number of reasons:
* Currently, thrust installs its config in a non-standard place, where it won't be found using the standard methods like `CMAKE_PREFIX_PATH` (NVIDIA/thrust#1292). Also, lots of pre-installed thrust's don't have a cmake installed with them in the first place (NVIDIA/thrust#1294)
* thrust, when used via `add_subdirectory()`/`CPMAddPackage` won't install itself, PR pending (NVIDIA/thrust#1297).
* In order to use the development version of thrust, once the above is merged, one might also hit a version mismatch, PR pending (NVIDIA/thrust#1298).

I'm kinda ambivalent about what the best way to proceed is. One could choose option 1 for now and switch to doing things properly once a new version of thrust is released (or once the PRs above are merged, depend on a specific version from thrust's `main` branch). Or do option 2 correctly already, which means that things will only really become usable as thrust gets updated.

FWIW, this is the diff for option 2:
```diff
diff --git a/cmake/RMM_thirdparty.cmake b/cmake/RMM_thirdparty.cmake
index 319d031..ab5abbc 100644
--- a/cmake/RMM_thirdparty.cmake
+++ b/cmake/RMM_thirdparty.cmake
@@ -23,6 +23,10 @@ CPMFindPackage(
   GIT_TAG 1.10.0
   VERSION 1.10.0
   GIT_SHALLOW TRUE
+  OPTIONS
+    # If there is no pre-installed thrust we can use, we'll install our fetched copy
+    # together with RMM
+    "THRUST_INSTALL TRUE"
   )

 thrust_create_target(rmm::Thrust FROM_OPTIONS)
diff --git a/rmm-config.cmake.in b/rmm-config.cmake.in
index a720acf..617828f 100644
--- a/rmm-config.cmake.in
+++ b/rmm-config.cmake.in
@@ -6,6 +6,8 @@ cmake_minimum_required(VERSION 3.17)
 include(CMakeFindDependencyMacro)
 find_dependency(CUDAToolkit)
 find_dependency(spdlog 1.7.0)
+find_dependency(Thrust 1.10.0)
+thrust_create_target(rmm::Thrust FROM_OPTIONS)

 include("${CMAKE_CURRENT_LIST_DIR}/rmm-targets.cmake")
 include("${CMAKE_CURRENT_LIST_DIR}/rmm-config-version.cmake")
```

Option 1 would be simply going to back to using `$<BUILD_INTERFACE:...>` for the dependency on `rmm::Thrust`, so that it doesn't get recorded in the installed package.

**Comment by kkraus14** [comment]
I'm in favor of option 1 with the caveat that we should present a warning if the found Thrust is lower than the version we specify, i.e. 1.10.0. This way users get information that they're going down an untested / unofficial path and they could choose to install a newer Thrust themselves for now.

**Comment by germasch** [comment]
I ended up doing a combined option 1, 2, and followed @kkraus' suggestion.

That is, we properly record the dependency on `rmm::Thrust`, and search for this dependency when used downstream. However, instead of failing if Thrust >= 1.10.0 can't be find, we print a big fat warning and create a fake (empty) `rmm::Thrust` target, which means that things will still work, falling back to the Thrust included with the CUDA toolkit.

File: rmm-config.cmake.in
Comment (kkraus14) [inline rmm-config.cmake.in:35]: Can we tweak this message to give more information about what thrust version was found and where it was found? i.e.
```
  message(WARNING "RMM could not find a Thrust library >= 1.10.0. Using Thrust version ${Thrust_VERSION} found at ${Thrust_PATH}.")
```

File: rmm-config.cmake.in
Comment (kkraus14) [inline rmm-config.cmake.in:38]: CMake error messages sometimes refer to the name underneath the ALIAS target and `missing_Thrust` is a bit misleading.
```suggestion
  add_library(found_Thrust INTERFACE)
  add_library(rmm::Thrust ALIAS found_Thrust)
```

File: rmm-config.cmake.in
Comment (germasch) [reply rmm-config.cmake.in:22]: Nope, I don't think so. I copy&pasted it from the actual `find_dependency`. Some grepping later, I think it's only used in `FeatureSummary()`, to decide whether to print info about the package.

File: rmm-config.cmake.in
Comment (germasch) [reply rmm-config.cmake.in:35]: Not easily, since we haven't found a thrust at this point.

File: rmm-config.cmake.in
Comment (germasch) [reply rmm-config.cmake.in:38]: Well, that was kinda the intention, since really no thrust was found in this case. I'll elaborate below.

**Comment by germasch** [comment]
So to clarify, I really went the cheap workaround way here: the installed RMM config will look for Thrust >= 1.10.0, and if it's not found (which is currently probably most of the time), it'll essentially just drop the dependency on Thrust (by making `rmm::Thrust` a dummy target that really doesn't do anything, other than exist.

Thrust only has been installing its own cmake config for a little while now (since 1.9.10), and that cmake config didn't make into CUDA releases or other Nvidia SDKs, I believe, and even when installed from source, it doesn't end up in the standard location. So in real life, it's not likely that an installed usable Thrust exists. This will work itself out over time.

But more immediately, RMM will pull in Thrust if no appropriate version is found (that is, pretty much always right now). Unfortunately, Thrust can't be installed by Thrust 1.10.0 when pulled in this way, but a Thrust PR is pending to fix that going forward. When Thrust 1.10.1 (or 1.11) comes out, everything should start working right. An intermediate solution might be for RMM to pull in a newer version of Thrust from its development branch, once those PRs are merged.

Right now there's no direct way to find a Thrust < 1.9.10 from cmake -- that's why I went the cheap way of effectively just ignoring the dependence on Thrust, since in real life there will be a Thrust in the CUDA toolkit and hence found automatically.

What to do somewhat depend on what Thrust RMM wants to support. If it's considered okay to use RMM with an older version of Thrust, I can support that more correctly by including a find module that'll find older Thrusts (this exists, I'll just have to install that, too). But if that's not a supported combination, I think the current way which prints a warning and otherwise still works is good enough of a bandaid until we can pull in a Thrust that will install itself together with RMM, so that RMM will provide its own recent enough version of Thrust if none exists yet.

**Comment by germasch** [comment]
@harrism, I committed all of your suggested doc changes, btw -- thanks!

File: rmm-config.cmake.in
Comment (harrism) [reply rmm-config.cmake.in:35]: I also don't really like the hard-coded "1.10.0" in the message. Can you parameterize that based on the specified dependency version?

File: rmm-config.cmake.in
Comment (kkraus14) [reply rmm-config.cmake.in:38]: Got it, this makes a lot more sense now, resolving, thanks!

File: rmm-config.cmake.in
Comment (kkraus14) [reply rmm-config.cmake.in:22]: Got it, thanks for the explanation. Resolving.

File: rmm-config.cmake.in
Comment (kkraus14) [reply rmm-config.cmake.in:35]: So a few ideas here:

1. We can keep the error message as is but include at least the cuda toolkit version since we already had `find_dependency(CUDAToolkit)`. This would help for debugging purposes if people are reporting broken builds due to RMM downstream.
2. We could explicitly use the cudatoolkit paths to find the `include/thrust` folder and manually create the target based on that instead of creating an empty Thrust target.
3. We could write / use a `FindThrust` cmake module to search through the standard system locations as well as CUDA install locations.

Additionally, if we went with option 2 or 3, we could capture the version by parsing `include/thrust/version.h` with a basic regex to grab the version number and then present a warning that version x.y.z is being used while we only test >= 1.10.0.

I'm strongly against option 3 and lean towards option 2.

**Comment by harrism** [comment]
Just FYI, I think RMM should work fine with thrust 1.9.x. HOWEVER, libcudf needs Thrust 1.10+ I think and the most common use cases of RMM also use cuDF.

File: rmm-config.cmake.in
Comment (germasch) [reply rmm-config.cmake.in:35]: Given that @harrism says that RMM should work with thrust 1.9.x, I'm inclined to support it properly then. Essentially all of 1., 2., 3. are kinda similar in that they're trying to find thrust, anyway, and ideally extract the version. If that code exists anyway, it might as well be done inside of `FindThrust.cmake` to make the rmm-config not even uglier than it already is. In fact, things will look much cleaner, I'm pretty sure.

**Comment by germasch** [comment]
So given the timeframe for the next release, I'm inclined to address the last little hitch that I'm aware of in this PR as well (besides finishing the thrust issue):

Right now, RMM does this:
```cmake
if(USE_NVTX)
  message(STATUS "Using Nvidia Tools Extension")
else()
  target_compile_definitions(rmm INTERFACE NVTX_DISABLE)
endif(USE_NVTX)
```

By default, `USE_NVTX` is off, which adds `-DNVTX_DISABLE` to the flags, and as such, the installed cmake configuration picks that up and hence a downstream project using RMM may unexpectedly end up with its NVTX disabled.

So this flag is something that maybe should be set for the benchmarks, but not generally for RMM and all of its consumers. Right now, it really doesn't need to be set at all, since I believe no nvtx is used anywhere at all.

An easy solution would be to get rid of the above, since it isn't even being used. That would also allow to reintroduce it later as `RMM_USE_NVTX`, if desired. But I'm not sure it's really clear at this point in which fashion nvtx would best be handled, anyway.

**Comment by harrism** [comment]
You can remove the NVTX stuff. I have an open PR that I plan to update in the next release to add some NVTX regions and I can add a new flag there if necessary.

**Comment by harrism** [comment]
Note that while burndown means "no new PRs", we still have about a week during burndown to finish in progress PRs.

File: cmake/install/FindThrust.cmake
Comment (harrism) [inline cmake/install/FindThrust.cmake:7]: What is the license for this?

**Comment by germasch** [comment]
> Note that while burndown means "no new PRs", we still have about a week during burndown to finish in progress PRs.

Yeah, I understand, that's why I didn't want to come with a little follow-up PR later. The latest push now removes `USE_NVTX` for the time being, and it also uses variables for the versions of thirdparty dependencies (thrust and spdlog). 

There's some asymmetry in that we required  Thrust >= 1.10.0 at build time (and fetch if necessary), but accept Thrust >= 1.9.0 when using the installed RMM. I think that's reasonable for the time being, though. I didn't leave a warning if Thrust < 1.10 ends up being used, but I can easily add one back if desired. When cmake finds an old Thrust version it will show what version it is, and its location, so it's at least somewhat visible in any case:

```
-- Found Thrust: /opt/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-7.5.0/cuda-10.2.89-cdzvh6elxmxj3kela4cnnvrzpjy2ett4/include (found suitable version "1.9.7", minimum required is "1.9.0")
```

 RMM will prefer a newer Thrust, found by config mode, though, before falling back to the old way.

There's a bit of a similar issue to `USE_NVTX` with `PER_THREAD_DEFAULT_STREAM`. But at least it's off by default, so it won't usually interfere. The nice thing about a header-only library is that one doesn't have to make these choices at build time for the library, hence it also would be better it they were not made (and recorded in the config). I think it'd better to just turn PTDS on (optionally) for the tests/benchmarks. (That also means one could have the tests compiled twice, with and w/o, where one wants to test both cases.)

File: cmake/install/FindThrust.cmake
Comment (germasch) [reply cmake/install/FindThrust.cmake:7]: Yeah, that's a good point. I guess it's this: https://gitlab.kitware.com/vtk/vtk-m/-/blob/783867eeb05e0a6538f9c520af02c3615651b4ed/LICENSE.txt. It does say 

```
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

 * Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

 * Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the
   distribution.

 * Neither the name of Kitware nor the names of any contributors may
   be used to endorse or promote products derived from this software
   without specific prior written permission.

```
So it seems pretty permissive, but IANAL.

I can write a custom one from scratch if this causes issues. Chances are it would look very similar, though, since really, the only thing that is not boilerplate in there is the version parsing, and that comes from [Nvidia](https://github.com/NVIDIA/thrust/blob/main/thrust/cmake/README.md).

**Review by kkraus14 — APPROVED** [review]
Approved pending the license question

File: cmake/install/FindThrust.cmake
Comment (harrism) [reply cmake/install/FindThrust.cmake:7]: I think it's the cmake license, not VTK. This is the BSD 3-clause license.  That is one of the licenses we can easily use. We still have a process we need to follow to add a new piece of software. I will check if RAPIDS already has done this for CMake code.

Comment (harrism) [reply cmake/install/FindThrust.cmake:7]: https://gitlab.kitware.com/cmake/cmake (readme says BSD 3-clause)

Comment (harrism) [reply cmake/install/FindThrust.cmake:7]: OK, we have approval, but before we can merge we need the complete license for this file included (ideally *in this file*).

Comment (germasch) [reply cmake/install/FindThrust.cmake]: Ok, so I copied in the full text from the LICENSE.txt above.

**Comment by harrism** [comment]
I just noticed that you *modified* FindThrust.cmake, rather than just pulling it in without changes. That changes the process. I would have to start over with a much more involved legal process.  Any chance of using the original unmodified FindThrust? (Or getting Thrust to provide their own?)

**Comment by germasch** [comment]
> I just noticed that in an earlier commit you _modified_ FindThrust.cmake, rather than just pulling it in without changes. That changes the process. I would have to start over with a much more involved legal process. Any chance of using the original unmodified FindThrust? (Or getting Thrust to provide their own?)

I started over and implemented a `FindThrust.cmake` from scratch. This at least has the advantage that I could actually make it provide compatible behavior to the configs installed by thrust since 1.9.10, which simplifies the `rmm-config.cmake` quite a bit. The only three lines of code that aren't my own are from https://github.com/NVIDIA/thrust/blob/main/thrust/cmake/README.md, which mirror the version calculations done in `thrust/version.h`.

>  Or somehow not needing FindThrust? (We didn't need it previously...)

Well, a prior revision of this PR was like that, if a newer Thrust wasn't found by config, it'd just keep its fingers crossed that some other thrust would be found automatically at build time (which would usually work, since it's in the cuda toolkit). But I think what's there now is cleaner, and at some point in the future, when Thrust that comes with its own cmake config is commonly available, the `FindThrust.cmake` can just be removed and things will still work.

It's worth noting that this `FindThrust.cmake` is deliberately put into `cmake/install/`, where it won't be found or used during the RMM build at all. The only reason it's there is so that it's installed together with the rest of RMM, so that a downstream cmake project that uses RMM handles RMM's thrust dependency properly. This PR, for the most part adds a new feature (ie., some additional files get installed for downstream use) and keeps existing use almost untouched.

**Comment by harrism** [comment]
This is fantastic. Thanks very much for doing this. It's not that we couldn't use the VTK-m code, its just that I have to go through a different process with the lawyers if we plan to modify it.

BTW, if you ever want to discuss anything not directly related to a PR off of github, you can find us on the RAPIDS public slack linked from this page: https://rapids.ai/community.html

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
@kkraus14 you might want to have a look before we merge?

**Review by kkraus14 — APPROVED** [review]

**Pull request merged**
